p2p/book/book.go

```diff
@@ -0,0 +1,434 @@
+		if queue[i].Connected && !queue[j].Connected {
+			return true
+		}
+		return queue[i].Class < queue[j].Class
+	})
+	for _, addr := range queue {
+		b.queue.PushBack(addr)
+	}
+	return nil
+}
+
+type Stats struct {
+	Total     int
+	Connected int
+
+	Public  int
+	Private int
+
+	Stale   int
+	Learned int
+	Stable  int
+}
+
+func (b *Book) Stats() Stats {
+	b.mu.Lock()
+	defer b.mu.Unlock()
+	stats := Stats{Total: len(b.known)}
+	for _, addr := range b.known {
+		if addr.Connected {
+			stats.Connected++
+		}
+		if addr.bucket == public {
+			stats.Public++
+		} else {
+			stats.Private++
+		}
+		switch addr.Class {
+		case stale:
+			stats.Stale++
+		case learned:
+			stats.Learned++
+		case stable:
+			stats.Stable++
+		}
+	}
+	return stats
+}
+
+func (b *Book) iterShareable(src ID, bucket bucket) iterator {
+	b.rng.Shuffle(len(b.shareable), func(i, j int) {
+		b.shareable[i], b.shareable[j] = b.shareable[j], b.shareable[i]
+	})
+	i := 0
+	return func() Address {
+		for {
+			if i == len(b.shareable) {
+				return nil
+			}
+			rst := b.shareable[i]
+			if rst.Class <= stale {
+				copy(b.shareable[i:], b.shareable[i+1:])
+				b.shareable[len(b.shareable)-1] = nil
+				b.shareable = b.shareable[:len(b.shareable)-1]
+				rst.shareable = false
+			} else {
+				i++
+				if rst.bucket == bucket && rst.ID != src {
+					return rst.Raw.Address
+				}
+			}
+		}
+	}
+}
+
+func (b *Book) drainQueue() iterator {
+	return func() Address {
+		for {
+			if b.queue.Len() == 0 {
+				return nil
+			}
+			rst := b.queue.Remove(b.queue.Front())
+			if rst.(*addressInfo).Class == deleted {
+				continue
+			}
+			return rst.(*addressInfo).Raw.Address
+		}
+	}
+}
+
+type iterator func() Address
+
+func take(n int, next iterator) []Address {
+	rst := make([]Address, 0, n)
+	for addr := next(); addr != nil; addr = next() {
+		rst = append(rst, addr)
+		if len(rst) == cap(rst) {
+			return rst
+		}
+	}
+	return rst
+}
+
+func persist(known map[ID]*addressInfo, w io.Writer) error {
+	checksum := crc64.New(crc64.MakeTable(crc64.ISO))
+	encoder := json.NewEncoder(io.MultiWriter(w, checksum))
+	sorted := make([]ID, 0, len(known))
+	for addr := range known {
+		sorted = append(sorted, addr)
+	}
+	sort.Slice(sorted, func(i, j int) bool {
+		return sorted[i] < sorted[j]
+	})
+	for _, id := range sorted {
+		addr := known[id]
+		if err := encoder.Encode(addr); err != nil {
+			return fmt.Errorf("json encoder failure for obj (%v): %w", addr, err)
+		}
+	}
+	if _, err := fmt.Fprintf(w, "%s", strconv.FormatUint(checksum.Sum64(), 10)); err != nil {
+		return fmt.Errorf("write checksum: %v", err)
+	}
+	return nil
+}
+
+func recover(known map[ID]*addressInfo, r io.Reader) error {
+	checksum := crc64.New(crc64.MakeTable(crc64.ISO))
+	scanner := bufio.NewScanner(r)
+	stored := uint64(0)
+	for scanner.Scan() {
+		if len(scanner.Bytes()) == 0 {
+			return fmt.Errorf("corrupted data: empty lines are not expected")
+		}
+		if scanner.Bytes()[0] != '{' {
```

Comment:
the old code assumes the first 4 bytes is the CRC value. that seems cleaner to me.
what's the reason to take a different approach here?

Reply:
i overlooked it when reviewing the change, or maybe suggested that by mistake

appending checksum doesn't require to keep content in memory, so it is either better to add it as a suffix or separate file